File: _plugins/03_process_images.rb
require 'open3'
require 'fileutils'

# Hook into the post_write lifecycle
Jekyll::Hooks.register :site, :post_write do |site|
  # Fetch configuration values
  config = site.config['image_processing'] || {}
  image_dir = File.join(site.source, config['source_dir'] ) # Default to 'images'
  output_dir = File.join(site.dest, config['output_dir'] ) # Default to 'images-thumbnails'
  thumbnail_width = config['thumbnail_width'] || 200 # Default size
  thumbnail_height = config['thumbnail_height'] || 200 # Default size

  # Ensure the output directory exists
  FileUtils.mkdir_p(output_dir)

  # Path to the Node.js script
  node_script = File.join(site.source, '_plugins/03_process_images.js')

  # Command to execute the Node.js script
  command = "node #{node_script} #{image_dir} #{output_dir} #{thumbnail_width} #{thumbnail_height}"
  stdout, stderr, status = Open3.capture3(command)

  if status.success?
    Jekyll.logger.info("Image Processing Complete:", stdout.strip)
  else
    Jekyll.logger.error("Image Processing Error:", stderr.strip)
  end
end

File: _plugins/03_process_images.js
const fs = require("fs");
const path = require("path");
const sharp = require("sharp");

// Get source, output directories, and thumbnail size from arguments
const [sourceDir, outputDir, thumbnail_width, thumbnail_height] =
  process.argv.slice(2);

if (!sourceDir || !outputDir || !thumbnail_width || !thumbnail_height) {
  console.error(
    "Usage: node process_images.js <sourceDir> <outputDir> <thumbnail_width> <thumbnail_height>"
  );
  process.exit(1);
}

const width = parseInt(thumbnail_width, 10);
const height = parseInt(thumbnail_height, 10);

if (isNaN(height) || isNaN(width)) {
  console.error("Thumbnail size must be a valid number.");
  process.exit(1);
}

// Ensure the output directory exists
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

// Process images
fs.readdir(sourceDir, (err, files) => {
  if (err) {
    console.error("Error reading source directory:", err);
    process.exit(1);
  }

  console.log(`Processing width:${width} height:${height}`);

  files.forEach((file) => {
    const inputPath = path.join(sourceDir, file);
    const outputPath = path.join(outputDir, file);

    // Check file extension to process only images
    if (/\.(jpe?g|png|webp|tiff|gif)$/i.test(file)) {
      sharp(inputPath)
        .resize(width, height) // Use dynamic size
        .toFile(outputPath)
        .then(() => console.log(`Processed: ${file}`))
        .catch((err) => console.error(`Error processing ${file}:`, err));
    }
  });
});
